src/trace/ord/merge_batcher/mod.rs

```diff
@@ -270,51 +312,85 @@ impl<D: Ord, R: MonoidValue> MergeSorter<D, R> {
                     }
                 };
 
+                if cfg!(debug_assertions) {
+                    if let Some((last, _)) = result.last() {
+                        debug_assert!(last <= &merged.0);
+                    }
+                }
+
                 // Safety: We've checked that the current vec has spare capacity available
                 //         as part of the loop's condition
                 debug_assert!(result.has_spare_capacity());
                 unsafe { result.push_unchecked(merged) };
             }
-
-            if result.capacity() == result.len() {
-                output.push(result);
-                result = self.buffer();
-            }
+            debug_assert!(result.is_sorted_by(|(a, _), (b, _)| Some(a.cmp(b))));
 
             if head1.is_empty() {
-                if head1.capacity() == Self::BUFFER_ELEMENTS {
+                if head1.capacity() >= Self::BUFFER_ELEMENTS {
                     self.stash.push(Vec::from(head1));
                 }
 
-                head1 = list1.pop_front().map_or_else(VecDeque::new, VecDeque::from);
+                // TODO: We could also attempt to pop from list2 here
+                head1 = list1
+                    .pop()
+                    .or_else(|| list2.pop())
+                    .map_or_else(VecDeque::new, VecDeque::from);
             }
+
             if head2.is_empty() {
-                if head2.capacity() == Self::BUFFER_ELEMENTS {
+                if head2.capacity() >= Self::BUFFER_ELEMENTS {
                     self.stash.push(Vec::from(head2));
                 }
 
-                head2 = list2.pop_front().map_or_else(VecDeque::new, VecDeque::from);
+                // TODO: We could also attempt to pop from list1 here
+                head2 = list2
+                    .pop()
+                    // .or_else(|| list1.pop())
+                    .map_or_else(VecDeque::new, VecDeque::from);
+            }
+
+            // If result is empty we have to start a new result vec to make sure
```

Comment:
I don't think this is right.  The invariant that the original code appears to maintain is that not just each individual batch in `list1` and `list2`, but the concatenation of batches is sorted.  If that invariant holds, then it's totally fine to continue pushing to the same result vector.

Reply:
It's not, take this situation:
```
result = [1, 2, 3, 4]
list2 = [10]

// pop a new list1
list1 = [1]
```
The next iteration will push `1` to `result` which results in `[1, 2, 3, 4, 1]` which is incorrect

Reply:
So, I applied your suggested fix but I don't think it's correct, now the assertion that `debug_assert!(last <= &merged.0);` gets set off